### ISSUE TITLE
Intermittent failure in replcompletions test

### ISSUE THREAD

**Issue by tkelman** [JuliaLang/julia#10875]
```
ERROR: LoadError: LoadError: test failed: (0:-1 == 17:21)
 in expression: r == endof(s) - 4:endof(s)
 in anonymous at task.jl:1388
while loading replcompletions.jl, in expression starting on line 358
while loading /tmp/julia/share/julia/test/runtests.jl, in expression starting on line 3
```
from https://travis-ci.org/JuliaLang/julia/jobs/59003679, commit 3abc13e on 64 bit Linux

other occurrences:
http://buildbot.e.ip.saba.us:8010/builders/build_ubuntu12.04-x86/builds/1176/steps/shell_2/logs/stdio
http://buildbot.e.ip.saba.us:8010/builders/build_ubuntu14.04-x86/builds/1118/steps/shell_2/logs/stdio
https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.3896/job/riq6yjykne09kt02
http://buildbot.e.ip.saba.us:8010/builders/build_ubuntu14.04-x64/builds/1166/steps/shell_2/logs/stdio
https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.3906/job/vjlfoq2r2stpb3dn
https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.3917/job/egkj811th386i2lo
https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.3919/job/ows2ywtlbq36o2yd
https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.3920/job/oo3t5vlxqdns34a4
http://buildbot.e.ip.saba.us:8010/builders/build_ubuntu12.04-x86/builds/1178/steps/shell_2/logs/stdio
http://buildbot.e.ip.saba.us:8010/builders/build_ubuntu12.04-x86/builds/1181/steps/shell_2/logs/stdio
https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.3924/job/1av04y9419v0tjcc
https://travis-ci.org/JuliaLang/julia/jobs/59102646
https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.3939/job/4n4h7uyvyt2xdb7s
https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.3941/job/y8mcyorudtpb7x99

(I think you get the point)

**Comment by mbauman** [comment]
I first saw this when testing the tupopocalypse branch on my Mac on https://github.com/JuliaLang/julia/commit/c3bbbb3716e7aa2e068d54727c49c8659f57e85c, but I couldn't reproduce it.  Was that the first sighting?  Perhaps I should have said something then.

**Comment by tkelman** [comment]
Found a reduced set of tests for this (also on win32), and it looks like whatever's causing this is probably the same underlying bug as what's causing the other two new failures since the tuple merge. Type comparison to `Array{Any,1}`.

I added `println(Base.shell_parse(s,true))` between these two lines https://github.com/JuliaLang/julia/blob/a49ae5374a0a4475361d1f8cf42fe864e5e8daca/test/replcompletions.jl#L367-L368

and I get the following:
```
julia> Base.runtests(["core","lineedit","replcompletions"], 1)
     * core                 in  12.22 seconds
     * lineedit             in   2.84 seconds
     * replcompletions     exception on 1: ERROR: LoadError: TypeError: shell_parse: in typeassert, expected Array{Any,1}, got Array{Any,1}
 in shell_parse at string.jl:1174
 in anonymous at replcompletions.jl:368
 in open at iostream.jl:111
 in anonymous at replcompletions.jl:365
 in anonymous at no file:364
 in runtests at D:\cygwin64\home\Tony\julia32\usr\share\julia\test\testdefs.jl:77
 in anonymous at multi.jl:626
 in run_work_thunk at multi.jl:587
 in remotecall_fetch at multi.jl:675
 in anonymous at task.jl:1386
while loading replcompletions.jl, in expression starting on line 358
ERROR: LoadError: LoadError: TypeError: shell_parse: in typeassert, expected Array{Any,1}, got Array{Any,1}
 in shell_parse at string.jl:1174
 in anonymous at replcompletions.jl:368
 in open at iostream.jl:111
 in anonymous at replcompletions.jl:365
 in anonymous at no file:364
 in runtests at D:\cygwin64\home\Tony\julia32\usr\share\julia\test\testdefs.jl:77
 in anonymous at multi.jl:626
 in run_work_thunk at multi.jl:587
 in remotecall_fetch at multi.jl:675
 in anonymous at task.jl:1386
while loading replcompletions.jl, in expression starting on line 358
while loading D:\cygwin64\home\Tony\julia32\usr\share\julia\test\runtests.jl, in expression starting on line 3

ERROR: A test has failed. Please submit a bug report (https://github.com/JuliaLang/julia/issues)
including error messages above and the output of versioninfo():
Julia Version 0.4.0-dev+4388
Commit 2a81411* (2015-04-20 15:23 UTC)
Platform Info:
  System: Windows (i686-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-2630QM CPU @ 2.00GHz
  WORD_SIZE: 32
  BLAS: libopenblas (DYNAMIC_ARCH NO_AFFINITY Nehalem)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3

 in error at error.jl:19
 in runtests at interactiveutil.jl:400
```

@JeffBezanson I wonder if maybe SimpleVector is getting mixed up with `Array{Any,1}` internally somehow?

**Comment by JeffBezanson** [comment]
7ec501d6e85aa1614c3a330bbbbf907be79bed37 is a possible fix.

**Comment by tkelman** [comment]
Apparently not, bummer https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.4054/job/6apo08twe807nxbl

**Comment by JeffBezanson** [comment]
The failures we've seen seem to indicate at least two bugs. @vtjnash & I fixed a definite bug in the type cache, but there have also been assertion failures that seem more like memory corruption. The failure described in this issue could be a third bug.

**Comment by JeffBezanson** [comment]
@vtjnash I checked the ordering of entries in Array's cache. Everything is fine on startup, and then after running core.jl there are many pairs of mis-ordered entries. I tried using only object_id and never dt->uid in typekey_compare and the problem went away. I'm not totally sure why this happens. I don't believe a uid ever changes.

**Comment by tkelman** [comment]
my 32 bit dockerfile running at commit 5e99f74 reproduces this on my machine, if there's anything you want me to try

**Comment by tkelman** [comment]
let's call this tentatively fixed by b993658e1699db110ac85446a8fb3c251e65476a and reopen if it happens again
